File: go/2016/day04/day04.go
package main

import (
	"flag"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type room struct {
	name       string
	roomNumber int
	checksum   string
	occurrence map[rune]int
}

type freqEntry struct {
	char  rune
	count int
}

type SortEntryByCountAndAlpha []freqEntry

func (a SortEntryByCountAndAlpha) Len() int      { return len(a) }
func (a SortEntryByCountAndAlpha) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
func (a SortEntryByCountAndAlpha) Less(i, j int) bool {
	if a[i].count == a[j].count {
		return a[i].char < a[j].char
	}
	return a[i].count > a[j].count
}

func (r *room) getObservedChars(size int) []freqEntry {
	occurrence := r.occurrence
	observed := make([]freqEntry, 5)
	for k, v := range occurrence {
		entry := freqEntry{k, v}
		observed = append(observed, entry)
		sort.Sort(SortEntryByCountAndAlpha(observed))
		observed = observed[0:5]
	}
	return observed
}

func (r *room) decryptRoomName() int {
	name := r.name
	roomNumber := r.roomNumber

	var sb strings.Builder

	for _, c := range name {
		if c == '-' {
			sb.WriteRune(' ')
		}
		sb.WriteRune(rune('a' + (int(c)-'a'+roomNumber)%26))
		if sb.String() == "northpole" {
			return roomNumber
		}
	}
	return -1
}

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)

	if err != nil {
		panic(err)
	}

	input := string(bytes)

	fmt.Printf("AoC 2016, Day4, Part1 solution is: %d \n", part1(&input))
	fmt.Printf("AoC 2016, Day4, Part2 solution is: %d \n", part2(&input))
}

func part1(input *string) int {
	lines := strings.Split(*input, "\n")
	sum := 0

	for _, line := range lines {
		room := getRoom(line)
		freqEntries := room.getObservedChars(5)
		inputChecksum := room.checksum
		id := room.roomNumber

		var sb strings.Builder
		for _, fq := range freqEntries {
			sb.WriteRune(fq.char)
		}

		if sb.String() == inputChecksum {
			sum += id
		}
	}

	return sum
}

func part2(input *string) int {
	lines := strings.Split(*input, "\n")
	for _, line := range lines {
		room := getRoom(line)
		nr := room.decryptRoomName()
		if (nr) > -1 {
			return nr
		}
	}
	return -1
}

func getRoom(row string) room {
	re := regexp.MustCompile("([a-z-]+)-([0-9]+)\\[([a-z]+)\\]")

	parsed := re.FindStringSubmatch(row)

	if len(parsed) != 4 {
		fmt.Printf("Invalid row: %s \n", row)
	}

	name := parsed[1]
	id, _ := strconv.Atoi(parsed[2])
	check := parsed[3]

	occurrence := make(map[rune]int)

	for _, c := range name {
		if c == '-' {
			continue
		}
		occurrence[c] += 1
	}

	return room{
		name:       name,
		roomNumber: id,
		occurrence: occurrence,
		checksum:   check,
	}
}

File: python/2016/day04/day04.py
import re

pattern = r"^(.*)-(\d+)\[(.*)\]$"

with open("input.txt") as f:
    data_input = [line.strip() for line in f.readlines()]


def part1() -> int:
    count = 0
    for line in data_input:
        encrypted_name, sector, checksum = __extract_parts(line)

        occurrence = __build_occurrence_map(encrypted_name)
        sorted_keys = sorted(occurrence.items(), key=lambda item: (-item[1], item[0]))
        s = ""
        for c, _ in sorted_keys[0:len(checksum)]:
            s = s + c
        if s == checksum:
            count += sector

    return count


def part2() -> int:
    for line in data_input:
        encrypted_name, sector, checksum = __extract_parts(line)
        decrypted = []
        for char in encrypted_name:
            if char == '-':
                decrypted.append(' ')
            else:
                shifted = (ord(char) - ord('a') + sector) % 26
                decrypted.append(chr(ord('a') + shifted))
        decrypted_message = ''.join(decrypted)
        if decrypted_message == "northpole object storage":
            return sector

    return -1


def __build_occurrence_map(encrypted_name: str) -> dict:
    d = {}

    for c in encrypted_name:
        if c == "-":
            continue
        if c.isdigit():
            continue
        v = d.get(c, 0)
        d[c] = v + 1

    return d


def __extract_parts(line: str) -> (str, int, str):
    match = re.match(pattern, line)

    if match:
        return match.group(1), int(match.group(2)), match.group(3)


if __name__ == '__main__':
    print(f"Day04 solution for part1 is: {part1()}")
    print(f"Day04 solution for part2 is: {part2()}")
